CamOps Contract Market - Contract Generation

Algebro7 opened the pull request:
Opening this as a draft while I work on the next batch of functionality for the CamOps Contract Market. Any feedback welcome.

Algebro7 commented:
I'm flipping this for ready for review. Much of the work for this phase is now done--the final things involve implementing negotiation of terms and extracting the payment formula out of `Contract` to make it CamOps compliant. Both will require substantial refactoring so I'd rather get this merged first before I start breaking other things.

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/AtbMonthlyContractMarket.java:
```diff
@@ -473,7 +472,7 @@ public double calculatePaymentMultiplier(Campaign campaign, AtBContract contract
         }
 
-        multiplier *= contract.getContractType().getPaymentMultiplier();
+        multiplier *= contract.getContractType().getOperationsTempoMultiplier();
```
Thanks for including this. Will make implementing Fame & Infamy modifiers _much_ easier.

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/AtbMonthlyContractMarket.java:
```diff
@@ -498,6 +497,20 @@ public double calculatePaymentMultiplier(Campaign campaign, AtBContract contract
+            if (roll == 6) {
+                addFollowup(campaign, contract);
+                campaign.addReport(
+                    "Your employer has offered a follow-up contract (available on the <a href=\"CONTRACT_MARKET\">contract market</a>).");
```
It's really easy for these notices to get missed by the player. Can we either implement this as a dialog, or include selective color coding (and bolded font, for color-blind players)? There are plenty of examples of color-coded dialog in the Education Module, if you need something to base off of.

As a second point, this ideally needs to be added to a ResourceBundle, as per the style guide.

Algebro7 replied:
So I think those are good suggestions, but this is old code that I simply extracted to the market method and followup contracts aren't implemented in the CO version yet. I'm not opposed to doing it but it feels out of scope for this particular PR since I'm trying to change as little legacy AtB functionality as possible.

IllianiCBT replied:
Ah, didn't spot that! In that case feel free to ignore this part of the review. :)

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+package mekhq.campaign.market.contractMarket;
```
We need a legal statement here.

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/CamOpsContractMarket.java:
```diff
@@ -5,12 +5,12 @@
 import megamek.logging.MMLogger;
 import mekhq.MekHQ;
 import mekhq.campaign.Campaign;
```
We need a legal statement at the top of this file

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+    public int getSalvagePercentage(int roll) {
+        roll += salvageModifier;
+        if (roll < 4) {
+            return 0;
```
This would probably be better as a switch, combined with MathUtility.clamp to put roll within usable bounds without changing effective functionality.

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+
+    public int getSupportPercentage(int roll) {
+        roll += supportModifier;
+        int percentage = 0;
```
And here

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+    public int getTransportTerms(int roll) {
+        roll += transportModifier;
+        if (roll < 2) {
+            return 0;
```
And here

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+            salvageModifier += -1;
+            supportModifier += -1;
+        }
+        if (employer.isStingy()) {
```
Not super vital, but do we want the generosity modifiers to be stringed as if-else blocks, so that it's more obvious they're mutually exclusive?

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/ContractTerms.java:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    private void addUnitReputationModifiers(int reputationFactor) {
+        if (reputationFactor <= 0) {
```
I think this might benefit from the MathUtility.clam + switch combo I mentioned above.

IllianiCBT commented on MekHQ/src/mekhq/campaign/market/contractMarket/MissionSelector.java:
```diff
@@ -0,0 +1,142 @@
+package mekhq.campaign.market.contractMarket;
```
Legal statement

IllianiCBT reviewed (changes requested):
I've marked this as 'request changes' as I do have a couple points of feedback.

The switch comments may well come down to personal preference. I find switches a lot easier to read at a glance. So, if you disagree, feel free to disregard those points.

We definitely need to add the legal notices to the top of any files that are missing them -- this is one I need to get better about, too, which is the only reason I knew to check.

If possible, I would really like to see you adopt either a dialog notice or selectively color-coded (and bolded) text for the follow-up contract notice. It also probably should be in a ResourceDialog.

IllianiCBT commented:
I'm going to be away for a couple of days so won't be able to do a follow-up review until Monday. So long as the legal notices are added and we get some color coding on the follow-up notice, I'm happy for this to be treated as 'approved' for the purposes of merging.

Algebro7 commented:
@IllianiCBT Thanks a lot for the review, those were all great points. I didn't know we had `MathUtility.clamp()` so that made a big difference on simplifying those massive if/else blocks.

I believe I've addressed all of your points except for improvements on the legacy followup contract notification, which we can implement as a separate PR.

IllianiCBT commented:
Sorry for leaving you hanging, I completely forgot to approve this
